### ISSUE TITLE
Fortran 2008 submodules / build system changes

### ISSUE THREAD

**Issue by phil-blain** [CICE-Consortium/CICE#334]
I just want to open the discussion regarding the use of Fortran 2008 submodules in CICE. Some references for background:

- https://software.intel.com/en-us/blogs/2015/07/07/doctor-fortran-in-we-all-live-in-a-yellow-submodule
- https://software.intel.com/en-us/fortran-compiler-developer-guide-and-reference-submodule
- https://www.ibm.com/support/knowledgecenter/en/SSGH4D_16.1.0/com.ibm.xlf161.aix.doc/proguide/cfsubmodules.html
- http://fortranwiki.org/fortran/show/Submodules (dead link at the moment, [Google cache](http://webcache.googleusercontent.com/search?q=cache:T8vicSaOf8gJ:fortranwiki.org/fortran/show/Submodules+&cd=4&hl=fr&ct=clnk&gl=ca&client=ubuntu))
- https://software.intel.com/en-us/articles/fortran-2008-submodule-support-in-intel-fortran-compiler

From what I gather submodules are a great way to further modularize code, and also to separate module subroutines/function interfaces from their implementation. This brings some very useful improvements:
- Submodule 'Asub' of module 'A' can `use` module 'B' even if module 'B' `use`s module 'A'. (doing this with just modules is impossible because it creates a circular dependency.) This can lead to very flexible and more modular code design. 
- Incremental build time can be greatly reduced, because a change in a subroutine implementation (ex. 'Asub.F90' is modified), does not cause the files that `use` module A to be recompiled (since the interface did not change).

Note : compilation order is important, module 'A' needs to be compiled before its submodules.

The last bullet is of course dependent on the build system. At the moment the `makdep.c` program used in CICE to generate the makefile dependencies does not understand Fortran submodules. I think it could be added, but I also think that it would be a good opportunity to talk about maybe changing the build system to something more powerful that already has support for modern Fortran features (like submodules), such as CMake (or meson, but CMake has a much bigger user base). 

This would also mean that if we do choose to use submodules, CICE can only be built with compilers that support the feature. A table of supported compilers is here : 
http://fortranwiki.org/fortran/show/Fortran+2008+status (dead at the moment, [Google cache](http://webcache.googleusercontent.com/search?q=cache:3uwo_GtdQHcJ:fortranwiki.org/fortran/show/Fortran%2B2008%2Bstatus+&cd=2&hl=fr&ct=clnk&gl=ca&client=ubuntu))
Note : the table above shows most up-to-date compiler versions, not the minimum version that supports each feature.

What do you think?

**Comment by phil-blain** [comment]
I also found this blog post that shows how a makefile can be written to handle submodules : 
https://aoterodelaroza.github.io/devnotes/modern-fortran-makefiles/
This person wrote a dependency generator in awk.

**Comment by apcraig** [comment]
We have added a section to the Software Development Practice to summarize a policy regarding this issue, https://github.com/CICE-Consortium/About-Us/wiki/Software-Development-Practices, and decided we can close the issue at this point.  The policy generally states that we should avoid new features until we are confident they are robustly implemented in all typically installed compilers unless there is a good reason.

**Comment by dabail10** [comment]
I know we closed this, but I just talked to the CESM software engineers about this. Basically, we are at 2003 at this point. It is robustly supported by the main compilers that we routinely test with (intel, nag, gnu, and pgi). There have been a few snags with some f03 features not supported properly or buggy, but in general it is mostly robust. We won't be moving to 2008 until that is fully supported everywhere. I also think it is interesting that the code still has the F90 extension everywhere. Basically, there is no need to keep updating to F00, F03, etc. Everything after F90 is supported in the compilers at present.
